trieste/acquisition/function/greedy.py

```diff
@@ -349,3 +353,127 @@ def __call__(self, x: TensorType) -> TensorType:
         penalization = ((pairwise_distances / (self._radius + self._scale)) ** p + 1) ** (1 / p)
         return tf.reduce_prod(penalization, axis=-1)
+
+
+class FantasizeAcquisitionFunction(GreedyAcquisitionFunctionBuilder):
+    r"""
+    Builder of the acquisition function maker for greedily collecting batches, following the
+    kriging believer heuristic.
+
+    TODO: add stuff
+
+    Kriging believer allows us to perform batch Bayesian optimization with a standard (non-batch)
+    acquisition function.
+    By iteratively building a batch of points though sequentially maximizing
+    this acquisition function,
+    TODO: add stuff
+    it provides diverse batches of candidate points.
+    """
+
+    def __init__(
+        self,
+        acquisition_function_builder: AcquisitionFunctionBuilder = None,
+        fantasize_method: str = "KB",
+    ):
+        """
+        :param acquisition_function_builder:
+        :raise tf.errors.InvalidArgumentError: If ``fantasize_method`` is not "KB" or "sample".
+        """
+        if acquisition_function_builder is None:
+            self._builder = ExpectedImprovement()
+        else:
+            self._builder = acquisition_function_builder
+
+        # TODO assert fantasize mode
+        self._fantasize_method = fantasize_method
+
+    def prepare_acquisition_function(
+        self,
+        models: Mapping[str, ProbabilisticModel],
+        datasets: Optional[Mapping[str, Dataset]] = None,
+        pending_points: Optional[TensorType] = None,
+    ) -> AcquisitionFunction:
+        """
+        :param models: The models over each tag.
+        :param datasets: The data from the observer (optional).
+        :param pending_points: Points already chosen to be in the current batch (of shape [M,D]),
+            where M is the number of pending points and D is the search space dimension.
+        :return: An acquisition function.
+        """
+        if pending_points is not None:
+            fantasized_data = {
+                tag: _generate_fantasized_data(model, pending_points, self._fantasize_method)
+                for tag, model in models.items()
+            }
+            models = {
+                tag: FantasizedGPRModel(model, fantasized_data[tag])
+                for tag, model in models.items()
+            }
+
+        return self._builder.prepare_acquisition_function(models, datasets)
+
+
+def _generate_fantasized_data(model, pending_points, fantasize_method):
+    if fantasize_method == "KB":
+        fantasized_obs, _ = model.predict(pending_points)
+    elif fantasize_method == "sample":
+        fantasized_obs = model.sample(pending_points, num_samples=1)  # TODO check dim
+    else:
+        raise NotImplementedError(
+            f"Fantasize method not supported. "
+            f"Expected KB or sample, received "
+            f"{fantasize_method}"
+        )
+    return Dataset(pending_points, fantasized_obs)
+
+
+class FantasizedGPRModel(ProbabilisticModel):
```

Comment:
shouldn't this go to the models/gpflow folder?

Reply:
Good question. This is the only place where it's used, and it' not really a model, just a wrapper to allow KB, so to me it is less confusing to have it here. But happy to reconsider.

Reply:
looking at this, I'm not sure we need it at all - why not simply calling conditional predict and sample above? since we require the `FastUpdateModel` these methods would be expected to be there 
(although we should still check that they actually exist at runtime)

Reply:
```suggestion
class FantasizedGPRModel(FastUpdateModel):
```